swap/protocol_test.go

```diff
@@ -44,68 +46,174 @@ func init() {
 	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(colorable.NewColorableStderr(), log.TerminalFormat(true))))
 }
 
-/*
-TestHandshake creates two mock nodes and initiates an exchange;
-it expects a handshake to take place between the two nodes
-(the handshake would fail because we don't actually use real nodes here)
-*/
-func TestHandshake(t *testing.T) {
-	var err error
+// protocol tester based on a swap instance
+type swapTester struct {
+	*p2ptest.ProtocolTester
+	swap *Swap
+}
 
-	// setup test swap object
+// creates a new protocol tester for swap with a deployed chequebook
+func newSwapTester(t *testing.T) (*swapTester, func(), error) {
 	swap, clean := newTestSwap(t, ownerKey, nil)
-	defer clean()
 
-	ctx := context.Background()
-	err = testDeploy(ctx, swap)
+	err := testDeploy(context.Background(), swap)
 	if err != nil {
-		t.Fatal(err)
+		return nil, nil, err
 	}
+
 	// setup the protocolTester, which will allow protocol testing by sending messages
-	protocolTester := p2ptest.NewProtocolTester(swap.owner.privateKey, 2, swap.run)
+	protocolTester := p2ptest.NewProtocolTester(swap.owner.privateKey, 1, swap.run)
+	return &swapTester{
+		ProtocolTester: protocolTester,
+		swap:           swap,
+	}, clean, nil
+}
+
+// creates a test exchange for the handshakes
+func HandshakeMsgExchange(lhs, rhs *HandshakeMsg, id enode.ID) []p2ptest.Exchange {
+	return []p2ptest.Exchange{
+		{
+			Expects: []p2ptest.Expect{
+				{
+					Code: 0,
+					Msg:  lhs,
+					Peer: id,
+				},
+			},
+		},
+		{
+			Triggers: []p2ptest.Trigger{
+				{
+					Code: 0,
+					Msg:  rhs,
+					Peer: id,
+				},
+			},
+		},
+	}
+}
 
-	// shortcut to creditor node
-	debitor := protocolTester.Nodes[0]
-	creditor := protocolTester.Nodes[1]
+// helper function for testing the handshake
+// lhs is the HandshakeMsg we expect to be sent, rhs the one we receive
+// disconnects is a list of disconnect events to be expected
+func (s *swapTester) testHandshake(lhs, rhs *HandshakeMsg, disconnects ...*p2ptest.Disconnect) error {
+	if err := s.TestExchanges(HandshakeMsgExchange(lhs, rhs, s.Nodes[0].ID())...); err != nil {
+		return err
+	}
 
-	// set balance artifially
-	swap.saveBalance(creditor.ID(), -42)
+	if len(disconnects) > 0 {
+		return s.TestDisconnected(disconnects...)
+	}
 
-	// create the expected cheque to be received
-	cheque := newTestCheque()
+	// If we don't expect disconnect, ensure peers remain connected
+	err := s.TestDisconnected(&p2ptest.Disconnect{
+		Peer:  s.Nodes[0].ID(),
+		Error: nil,
+	})
 
-	// sign the cheque
-	cheque.Signature, err = cheque.Sign(swap.owner.privateKey)
+	if err == nil {
```

Comment:
? 

Isn't a `TestDisconnected` returning an `err` if it does not find the expected disconnect? So wouldn't here `err` be `nil` if it actually found the provided `Disconnect`, which is saying stay connected, which is what you wanted to test?

Reply:
Yes, on a disconnect `err` would be nil. Which is an error because because this is where we we want to make sure that there is no disconnect. (This function is copied from the network tests btw).

Reply:
It's a bit counter-intuitive to pass a `Disconnect` object when not expecting a disconnect...Can't this just be called with an empty object, which would signal that we don't expect a disconnect?

Then `err` would return `nil` only if there would be no disconnect.

If that's not possible, I guess that's the way to do it and we keep it (I understand you copied it from elsewhere)

Reply:
Afaik there is no other way to check if the node is disconnected or not. (It also would not be possible, if we passed `nil`, then the function would not know which peer to check).

I would keep the function as it is, so that we remain consistent with the handshake test style from the other protocols.